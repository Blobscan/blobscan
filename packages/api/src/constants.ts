export const DEFAULT_LIMIT = 50;
