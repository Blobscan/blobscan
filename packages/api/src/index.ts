import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
