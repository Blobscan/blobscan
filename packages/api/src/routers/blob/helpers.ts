import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
