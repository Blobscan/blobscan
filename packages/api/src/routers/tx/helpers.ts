import type { Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseTransactionSchema,
  baseBlobSchema,
  baseBlockSchema,
} from "../../zod-schemas";

const transactionSelect = {
  hash: true,
  fromId: true,
  toId: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  gasPrice: true,
  maxFeePerBlobGas: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  decodedFields: true,
  from: {
    select: {
      address: true,
      rollup: true,
    },
  },
} satisfies Prisma.TransactionSelect;

type PrismaTransaction = Prisma.TransactionGetPayload<{
  select: typeof transactionSelect;
}>;

export type CompletePrismaTransaction = Prettify<
  PrismaTransaction & {
    decodedFields: NonNullable<PrismaTransaction["decodedFields"]>;
    block: Prettify<
      Partial<ExpandedBlock> & {
        blobGasPrice: ExpandedBlock["blobGasPrice"];
      }
    >;
    blobs: Prettify<
      { blobHash: string } & {
        blob?: ExpandedBlob;
      }
    >[];
  }
>;
export function createTransactionSelect(expands: Expands) {
  const blockExpand = expands.block?.select;
  const blobExpand = expands.blob;

  return {
    ...transactionSelect,
    block: {
      select: {
        ...(blockExpand ?? {}),
        blobGasPrice: true,
      },
    },
    blobs: {
      select: {
        blobHash: true,
        ...(blobExpand ? { blob: blobExpand } : {}),
      },
      orderBy: {
        index: "asc",
      },
    },
  } satisfies Prisma.TransactionSelect;
}

export const responseTransactionSchema = baseTransactionSchema.extend({
  block: baseBlockSchema
    .omit({
      hash: true,
      number: true,
      timestamp: true,
    })
    .partial()
    .required({
      blobGasPrice: true,
    }),
  blobs: z.array(baseBlobSchema.partial().required({ versionedHash: true })),
});

export type ResponseTransaction = z.input<typeof responseTransactionSchema>;

export function toResponseTransaction(
  prismaTx: CompletePrismaTransaction
): ResponseTransaction {
  const { blobs: blobsOnTxs, block } = prismaTx;
  const normalizedFields = normalizePrismaTransactionFields(prismaTx);
  const derivedFields = deriveTransactionFields({
    ...prismaTx,
    blobGasPrice: block.blobGasPrice,
  });
  const blobs = blobsOnTxs.map(({ blobHash, blob }) =>
    blob
      ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
      : { versionedHash: blobHash }
  );

  return {
    ...prismaTx,
    ...normalizedFields,
    ...derivedFields,
    blobs,
  };
}
