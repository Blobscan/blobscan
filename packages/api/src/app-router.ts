import { z } from "@blobscan/zod";

import { publicProcedure } from "./procedures";
import { blobRouter } from "./routers/blob";
import { blobStoragesStateRouter } from "./routers/blob-storages-state";
import { blockRouter } from "./routers/block";
import { blockchainSyncStateRouter } from "./routers/blockchain-sync-state";
import { indexerRouter } from "./routers/indexer";
import { searchRouter } from "./routers/search";
import { statsRouter } from "./routers/stats";
import { transactionRouter } from "./routers/tx";
import { t } from "./trpc-client";

export const appRouter = t.router({
  block: blockRouter,
  tx: transactionRouter,
  blob: blobRouter,
  search: searchRouter,
  stats: statsRouter,
  indexer: indexerRouter,
  syncState: blockchainSyncStateRouter,
  blobStoragesState: blobStoragesStateRouter,
  healthcheck: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: "/healthcheck",
        summary: "connection healthcheck.",
        tags: ["system"],
      },
    })
    .input(z.void())
    .output(z.string())
    .query(() => "yay!"),
});

// export type definition of API
export type AppRouter = typeof appRouter;
