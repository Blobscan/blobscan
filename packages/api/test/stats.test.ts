import { beforeAll, describe, expect, it } from "vitest";

import dayjs from "@blobscan/dayjs";
import { omitDBTimestampFields } from "@blobscan/test";

import { appRouter } from "../src/app-router";
import type { TimeFrame } from "../src/middlewares/withTimeFrame";
import { createTestContext } from "./helpers";

const TIME_FRAMES: TimeFrame[] = ["1d", "7d", "15d", "30d", "180d", "360d"];

function runTimeFrameTests(
  name: string,
  statsFiller: () => Promise<unknown>,
  statsFetcher: (timeFrame: TimeFrame) => Promise<unknown>
) {
  return describe(name, () => {
    describe("when getting stats for a specific timeframe", () => {
      TIME_FRAMES.forEach((timeFrame) => {
        it(`should get daily stats for ${timeFrame}`, async () => {
          await statsFiller();

          const result = await statsFetcher(timeFrame);

          expect(result).toMatchSnapshot();
        });
      });
    });
  });
}

describe("Stats router", async () => {
  const systemDate = new Date("2023-09-01");
  const to = dayjs(systemDate).endOf("day").toISOString();
  let caller: ReturnType<typeof appRouter.createCaller>;
  let prisma: Awaited<ReturnType<typeof createTestContext>>["prisma"];

  beforeAll(async () => {
    const ctx = await createTestContext();

    prisma = ctx.prisma;
    caller = appRouter.createCaller(ctx);
  });

  describe("getBlobOverallStats", () => {
    it("should return the correct overall stats", async () => {
      await prisma.blobOverallStats.increment({
        from: 0,
        to: 9999,
      });

      const blobOverallStats = await caller.stats.getBlobOverallStats();

      expect(omitDBTimestampFields(blobOverallStats)).toMatchInlineSnapshot(`
        {
          "avgBlobSize": 1137.5,
          "totalBlobSize": "36400",
          "totalBlobs": 32,
          "totalUniqueBlobs": 7,
        }
      `);
    });
  });

  runTimeFrameTests(
    "getBlobDailyStats",
    async () => {
      await prisma.blobDailyStats.populate({
        to,
      });
    },

    (timeFrame) => caller.stats.getBlobDailyStats({ timeFrame })
  );

  describe("getBlockOverallStats", () => {
    it("should return the correct overall stats", async () => {
      await prisma.blockOverallStats.increment({
        from: 0,
        to: 9999,
      });

      const blockOverallStats = await caller.stats.getBlockOverallStats();

      expect(omitDBTimestampFields(blockOverallStats)).toMatchInlineSnapshot(`
        {
          "avgBlobAsCalldataFee": 10450145,
          "avgBlobFee": 105412688,
          "avgBlobGasPrice": 21.75,
          "totalBlobAsCalldataFee": "83601160",
          "totalBlobAsCalldataGasUsed": "3822780",
          "totalBlobFee": "843301504",
          "totalBlobGasUsed": "38786432",
          "totalBlocks": 8,
        }
      `);
    });
  });

  runTimeFrameTests(
    "getBlockDailyStats",
    () =>
      prisma.blockDailyStats.populate({
        to,
      }),
    (timeFrame) => caller.stats.getBlockDailyStats({ timeFrame })
  );

  describe("getTransactionOverallStats", () => {
    it("should return the correct overall stats", async () => {
      await prisma.transactionOverallStats.increment({
        from: 0,
        to: 9999,
      });
      const result = await caller.stats.getTransactionOverallStats();

      expect(omitDBTimestampFields(result)).toMatchInlineSnapshot(`
        {
          "avgMaxBlobGasFee": 101.875,
          "totalTransactions": 16,
          "totalUniqueReceivers": 4,
          "totalUniqueSenders": 4,
        }
      `);
    });
  });

  runTimeFrameTests(
    "getTransactionDailyStats",
    () => prisma.transactionDailyStats.populate({ to }),
    (timeFrame) => caller.stats.getTransactionDailyStats({ timeFrame })
  );
});
