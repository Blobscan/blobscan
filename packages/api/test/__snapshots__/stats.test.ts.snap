// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stats router > getBlobDailyStats > when getting stats for a specific timeframe > should get daily stats for 1d 1`] = `
{
  "avgBlobSizes": [
    1280,
  ],
  "days": [
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobSizes": [
    6400,
  ],
  "totalBlobs": [
    5,
  ],
  "totalUniqueBlobs": [
    3,
  ],
}
`;

exports[`Stats router > getBlobDailyStats > when getting stats for a specific timeframe > should get daily stats for 7d 1`] = `
{
  "avgBlobSizes": [
    1000,
    1280,
  ],
  "days": [
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobSizes": [
    4000,
    6400,
  ],
  "totalBlobs": [
    4,
    5,
  ],
  "totalUniqueBlobs": [
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlobDailyStats > when getting stats for a specific timeframe > should get daily stats for 15d 1`] = `
{
  "avgBlobSizes": [
    1180,
    1000,
    1280,
  ],
  "days": [
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobSizes": [
    5900,
    4000,
    6400,
  ],
  "totalBlobs": [
    5,
    4,
    5,
  ],
  "totalUniqueBlobs": [
    4,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlobDailyStats > when getting stats for a specific timeframe > should get daily stats for 30d 1`] = `
{
  "avgBlobSizes": [
    1033.333333333333,
    1180,
    1000,
    1280,
  ],
  "days": [
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobSizes": [
    6200,
    5900,
    4000,
    6400,
  ],
  "totalBlobs": [
    6,
    5,
    4,
    5,
  ],
  "totalUniqueBlobs": [
    2,
    4,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlobDailyStats > when getting stats for a specific timeframe > should get daily stats for 180d 1`] = `
{
  "avgBlobSizes": [
    1250,
    1033.333333333333,
    1180,
    1000,
    1280,
  ],
  "days": [
    "2023-05-10T00:00:00.000Z",
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobSizes": [
    7500,
    6200,
    5900,
    4000,
    6400,
  ],
  "totalBlobs": [
    6,
    6,
    5,
    4,
    5,
  ],
  "totalUniqueBlobs": [
    6,
    2,
    4,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlobDailyStats > when getting stats for a specific timeframe > should get daily stats for 360d 1`] = `
{
  "avgBlobSizes": [
    1066.666666666667,
    1250,
    1033.333333333333,
    1180,
    1000,
    1280,
  ],
  "days": [
    "2022-10-16T00:00:00.000Z",
    "2023-05-10T00:00:00.000Z",
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobSizes": [
    6400,
    7500,
    6200,
    5900,
    4000,
    6400,
  ],
  "totalBlobs": [
    6,
    6,
    6,
    5,
    4,
    5,
  ],
  "totalUniqueBlobs": [
    3,
    6,
    2,
    4,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlockDailyStats > when getting stats for a specific timeframe > should get daily stats for 1d 1`] = `
{
  "avgBlobAsCalldataFees": [
    5610000,
  ],
  "avgBlobFees": [
    121000000,
  ],
  "avgBlobGasPrices": [
    22,
  ],
  "days": [
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobAsCalldataFees": [
    "16830000",
  ],
  "totalBlobAsCalldataGasUsed": [
    "765000",
  ],
  "totalBlobFees": [
    "363000000",
  ],
  "totalBlobGasUsed": [
    "16500000",
  ],
  "totalBlocks": [
    3,
  ],
}
`;

exports[`Stats router > getBlockDailyStats > when getting stats for a specific timeframe > should get daily stats for 7d 1`] = `
{
  "avgBlobAsCalldataFees": [
    5000000,
    5610000,
  ],
  "avgBlobFees": [
    100000000,
    121000000,
  ],
  "avgBlobGasPrices": [
    20,
    22,
  ],
  "days": [
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobAsCalldataFees": [
    "5000000",
    "16830000",
  ],
  "totalBlobAsCalldataGasUsed": [
    "250000",
    "765000",
  ],
  "totalBlobFees": [
    "100000000",
    "363000000",
  ],
  "totalBlobGasUsed": [
    "5000000",
    "16500000",
  ],
  "totalBlocks": [
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlockDailyStats > when getting stats for a specific timeframe > should get daily stats for 15d 1`] = `
{
  "avgBlobAsCalldataFees": [
    5610000,
    5000000,
    5610000,
  ],
  "avgBlobFees": [
    121000000,
    100000000,
    121000000,
  ],
  "avgBlobGasPrices": [
    22,
    20,
    22,
  ],
  "days": [
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobAsCalldataFees": [
    "5610000",
    "5000000",
    "16830000",
  ],
  "totalBlobAsCalldataGasUsed": [
    "255000",
    "250000",
    "765000",
  ],
  "totalBlobFees": [
    "121000000",
    "100000000",
    "363000000",
  ],
  "totalBlobGasUsed": [
    "5500000",
    "5000000",
    "16500000",
  ],
  "totalBlocks": [
    1,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlockDailyStats > when getting stats for a specific timeframe > should get daily stats for 30d 1`] = `
{
  "avgBlobAsCalldataFees": [
    5610000,
    5610000,
    5000000,
    5610000,
  ],
  "avgBlobFees": [
    121000000,
    121000000,
    100000000,
    121000000,
  ],
  "avgBlobGasPrices": [
    22,
    22,
    20,
    22,
  ],
  "days": [
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobAsCalldataFees": [
    "5610000",
    "5610000",
    "5000000",
    "16830000",
  ],
  "totalBlobAsCalldataGasUsed": [
    "255000",
    "255000",
    "250000",
    "765000",
  ],
  "totalBlobFees": [
    "121000000",
    "121000000",
    "100000000",
    "363000000",
  ],
  "totalBlobGasUsed": [
    "5500000",
    "5500000",
    "5000000",
    "16500000",
  ],
  "totalBlocks": [
    1,
    1,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlockDailyStats > when getting stats for a specific timeframe > should get daily stats for 180d 1`] = `
{
  "avgBlobAsCalldataFees": [
    5610000,
    5610000,
    5610000,
    5000000,
    5610000,
  ],
  "avgBlobFees": [
    121000000,
    121000000,
    121000000,
    100000000,
    121000000,
  ],
  "avgBlobGasPrices": [
    22,
    22,
    22,
    20,
    22,
  ],
  "days": [
    "2023-05-10T00:00:00.000Z",
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobAsCalldataFees": [
    "5610000",
    "5610000",
    "5610000",
    "5000000",
    "16830000",
  ],
  "totalBlobAsCalldataGasUsed": [
    "255000",
    "255000",
    "255000",
    "250000",
    "765000",
  ],
  "totalBlobFees": [
    "121000000",
    "121000000",
    "121000000",
    "100000000",
    "363000000",
  ],
  "totalBlobGasUsed": [
    "5500000",
    "5500000",
    "5500000",
    "5000000",
    "16500000",
  ],
  "totalBlocks": [
    1,
    1,
    1,
    1,
    3,
  ],
}
`;

exports[`Stats router > getBlockDailyStats > when getting stats for a specific timeframe > should get daily stats for 360d 1`] = `
{
  "avgBlobAsCalldataFees": [
    44941160,
    5610000,
    5610000,
    5610000,
    5000000,
    5610000,
  ],
  "avgBlobFees": [
    17301504,
    121000000,
    121000000,
    121000000,
    100000000,
    121000000,
  ],
  "avgBlobGasPrices": [
    22,
    22,
    22,
    22,
    20,
    22,
  ],
  "days": [
    "2022-10-16T00:00:00.000Z",
    "2023-05-10T00:00:00.000Z",
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalBlobAsCalldataFees": [
    "44941160",
    "5610000",
    "5610000",
    "5610000",
    "5000000",
    "16830000",
  ],
  "totalBlobAsCalldataGasUsed": [
    "2042780",
    "255000",
    "255000",
    "255000",
    "250000",
    "765000",
  ],
  "totalBlobFees": [
    "17301504",
    "121000000",
    "121000000",
    "121000000",
    "100000000",
    "363000000",
  ],
  "totalBlobGasUsed": [
    "786432",
    "5500000",
    "5500000",
    "5500000",
    "5000000",
    "16500000",
  ],
  "totalBlocks": [
    1,
    1,
    1,
    1,
    1,
    3,
  ],
}
`;

exports[`Stats router > getTransactionDailyStats > when getting stats for a specific timeframe > should get daily stats for 1d 1`] = `
{
  "avgMaxBlobGasFees": [
    100,
  ],
  "days": [
    "2023-08-31T00:00:00.000Z",
  ],
  "totalTransactions": [
    3,
  ],
  "totalUniqueReceivers": [
    2,
  ],
  "totalUniqueSenders": [
    2,
  ],
}
`;

exports[`Stats router > getTransactionDailyStats > when getting stats for a specific timeframe > should get daily stats for 7d 1`] = `
{
  "avgMaxBlobGasFees": [
    100,
    100,
  ],
  "days": [
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalTransactions": [
    2,
    3,
  ],
  "totalUniqueReceivers": [
    1,
    2,
  ],
  "totalUniqueSenders": [
    2,
    2,
  ],
}
`;

exports[`Stats router > getTransactionDailyStats > when getting stats for a specific timeframe > should get daily stats for 15d 1`] = `
{
  "avgMaxBlobGasFees": [
    100,
    100,
    100,
  ],
  "days": [
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalTransactions": [
    5,
    2,
    3,
  ],
  "totalUniqueReceivers": [
    3,
    1,
    2,
  ],
  "totalUniqueSenders": [
    4,
    2,
    2,
  ],
}
`;

exports[`Stats router > getTransactionDailyStats > when getting stats for a specific timeframe > should get daily stats for 30d 1`] = `
{
  "avgMaxBlobGasFees": [
    100,
    100,
    100,
    100,
  ],
  "days": [
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalTransactions": [
    2,
    5,
    2,
    3,
  ],
  "totalUniqueReceivers": [
    1,
    3,
    1,
    2,
  ],
  "totalUniqueSenders": [
    1,
    4,
    2,
    2,
  ],
}
`;

exports[`Stats router > getTransactionDailyStats > when getting stats for a specific timeframe > should get daily stats for 180d 1`] = `
{
  "avgMaxBlobGasFees": [
    100,
    100,
    100,
    100,
    100,
  ],
  "days": [
    "2023-05-10T00:00:00.000Z",
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalTransactions": [
    1,
    2,
    5,
    2,
    3,
  ],
  "totalUniqueReceivers": [
    1,
    1,
    3,
    1,
    2,
  ],
  "totalUniqueSenders": [
    1,
    1,
    4,
    2,
    2,
  ],
}
`;

exports[`Stats router > getTransactionDailyStats > when getting stats for a specific timeframe > should get daily stats for 360d 1`] = `
{
  "avgMaxBlobGasFees": [
    110,
    100,
    100,
    100,
    100,
    100,
  ],
  "days": [
    "2022-10-16T00:00:00.000Z",
    "2023-05-10T00:00:00.000Z",
    "2023-08-03T00:00:00.000Z",
    "2023-08-20T00:00:00.000Z",
    "2023-08-28T00:00:00.000Z",
    "2023-08-31T00:00:00.000Z",
  ],
  "totalTransactions": [
    3,
    1,
    2,
    5,
    2,
    3,
  ],
  "totalUniqueReceivers": [
    3,
    1,
    1,
    3,
    1,
    2,
  ],
  "totalUniqueSenders": [
    3,
    1,
    1,
    4,
    2,
    2,
  ],
}
`;
